pkg/specgen/generate/validate.go

```diff
@@ -11,20 +11,25 @@ import (
 	"github.com/containers/podman/v4/pkg/rootless"
 	"github.com/containers/podman/v4/pkg/specgen"
 	"github.com/containers/podman/v4/utils"
+	spec "github.com/opencontainers/runtime-spec/specs-go"
 )
 
 // Verify resource limits are sanely set when running on cgroup v1.
 func verifyContainerResourcesCgroupV1(s *specgen.SpecGenerator) ([]string, error) {
 	warnings := []string{}
+	resourceNil := &spec.LinuxResources{}
 
 	sysInfo := sysinfo.New(true)
 
-	if s.ResourceLimits != nil && rootless.IsRootless() {
-		s.ResourceLimits = nil
+	if s.ResourceLimits == nil {
+		s.ResourceLimits = &spec.LinuxResources{}
+	}
+
+	if rootless.IsRootless() && s.ResourceLimits != resourceNil {
 		warnings = append(warnings, "Resource limits are not supported and ignored on cgroups V1 rootless systems")
 	}
 
-	if s.ResourceLimits == nil {
+	if s.ResourceLimits == resourceNil {
```

Comment:
I find the code very hard to read because we're mixing the condition checks with setting `s.ResourceLimits` in case it's nil.

Can we rewrite it and `return []string{"Resource limits are not supported and ignored on cgroups V1 rootless systems"}, nil` immediately?

We should also move down the declarations of `sysInfo` and `warnings`.

Reply:
yes, please rewrite it because it is too difficult to follow now.

Just return the warning immediately, no need to set `s.ResourceLimits` to a dummy value